README.md
# DN2A - Digital Neural Networks Architecture #

[![Build Status](https://travis-ci.org/antoniodeluca/dn2a.js.svg?branch=master)](https://travis-ci.org/antoniodeluca/dn2a.js)
[![Code Climate](https://codeclimate.com/github/antoniodeluca/dn2a.js.png)](https://codeclimate.com/github/antoniodeluca/dn2a.js)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/antoniodeluca/dn2a.js/master/LICENSE)
[![npm version](https://badge.fury.io/js/dn2a.svg)](https://badge.fury.io/js/dn2a)

---

## About ##

DN2A is a set of highly decoupled JavaScript modules for Neural Networks and Artificial Intelligence development.

Each module is based on injection by configuration.

You can use a single module alone, more of them together or just the complete set.

DN2A main goal is to allow you to design, train and use without pain Single Neural Networks as well as very powerful Neural Networks Chains through which implement your Artificial Intelligence solution.

DN2A side goals are to simplify integration, to speed up training/querying, to allow clustering and to represent the architecture and the relative data of each Neural Network as a (re)combinable string strain that will be usable within genetics optimization techniques.

----------

## Features ##

- **Modularized components**: helps the development and the clear separation of concerns with great benefits for who wants to use mixed solutions.
- **Configurable precision**: helps to avoid the noise deriving from operation errors and default system precision limits with great improvement of the learning speed and performance stability.
- **Configuration checker**: helps to write less details about configuration and to keep compatibility with older version while the project evolves.
- **StepByStep training**: helps to train neural networks doing a single iteration over the passed information.
- **StopAtGoal training**: helps to train neural networks doing a finite number of iterations over the passed information until a specific parametric condition is reached.
- **Continuous training**: helps to train neural networks doing an infinite number of iterations over the passed information.
- TODO (Brain) **Data normalization**: helps to simplify the interaction within your real domain.
- TODO (Cerebrum) **Networks composition**: helps to create very effective architectures of multiple neural networks able to obtain advanced behaviours like in deep learning.
- TODO (Cerebrum) **Computation parallelization**: helps to improve the scalability of your whole system.
- TODO (Brain) **Sessions intercommunication**: helps to improve the scalability of your whole system.

----------

## Modules ##

### Neuron ###
Module able to facilitate the representation of the data structure around Neurons and to hold relative common functionalities.

### Synapse ###
Module able to facilitate the representation of the data structure around Synapses and to hold relative common functionalities.

### Network ###
Module, available in different variations, able to use Neurons and Synapses to implement configurable and autonomous Neural Networks.

#### Available Network Types ####

1.  **alpha**: standard feed forward neural network with error back propagation controlled by layer dimensions, learning mode, learning rate, momentum rate, maximum allowed error and maximum number of epochs.

### Cerebrum ###
Module for the management of multiple Neural Networks in terms of configuration/coordination, training/querying chaining and parallel computing.

### Brain ###
Module for the management of data normalization, integration/intercommunication with other external software and monitoring of the whole session.

----------

## Tutorials ##

### Using in Node  ###

To install the library through NPM:

	npm install dn2a

### Using in the Browser ###

To install the library through Bower:

	bower install dn2a

### Training and Querying a Single Network with default parametrization (ES5) ###

```javascript
// Importation
var DN2A = require("dn2a");

// Instantiation
var neuralNetwork = new DN2A.NetworkAlpha();

// Training
var trainingPatterns = [
    {
        input: [0, 0],
        output: [0]
    },
    {
        input: [0, 1],
        output: [1]
    },
    {
        input: [1, 0],
        output: [1]
    },
    {
        input: [1, 1],
        output: [0]
    }
];
neuralNetwork.train(trainingPatterns);

// Querying
var inputPatterns = [
    [0, 0],
    [0, 1],
    [1, 0],
    [1, 1]
];
neuralNetwork.query(inputPatterns, function(queryingStatus) {
    inputPatterns.forEach(function(inputPatten, inputPatternIndex) {
        console.log("[" + inputPatterns[inputPatternIndex].join(", ") + "] => [" + queryingStatus.outputPatterns[inputPatternIndex].join(", ") + "]");
    });
});
```

### Training and Querying a Single Network with custom parametrization (ES5) ###

```javascript
// Importation
var DN2A = require("dn2a");

// Instantiation
// The object expected by the constructor can specify properties that describe the neural network.
// The list of the valid properties together their accepted ranges and default values is reported in this README file.
// The object can be completely omitted and in this case default values are used for all properties.
var neuralNetwork = new DN2A.NetworkAlpha();

var neuralNetwork = new DN2A.NetworkAlpha({
    layerDimensions: [2, 4, 4, 1],
    learningMode: "continuous",
    learningRate: 0.3,
    momentumRate: 0.7,
    maximumError: 0.005,
    maximumEpoch: 20000,
    dataRepository: {},
    neuron: {
        generator: DN2A.Neuron
    },
    synapse: {
        generator: DN2A.Synapse
    },
    numbersPrecision: 32
});

// Training
var trainingPatterns = [
    {
        input: [0, 0],
        output: [0]
    },
    {
        input: [0, 1],
        output: [1]
    },
    {
        input: [1, 0],
        output: [1]
    },
    {
        input: [1, 1],
        output: [0]
    }
];
neuralNetwork.train(trainingPatterns);

// Querying
var inputPatterns = [
    [0, 0],
    [0, 1],
    [1, 0],
    [1, 1]
];
neuralNetwork.query(inputPatterns, function(queryingStatus) {
    inputPatterns.forEach(function(inputPatten, inputPatternIndex) {
        console.log("[" + inputPatterns[inputPatternIndex].join(", ") + "] => [" + queryingStatus.outputPatterns[inputPatternIndex].join(", ") + "]");
    });
});
```

### Training and Querying a Single Network with evolution feedback (ES5) ###

```javascript
// Importation
var DN2A = require("dn2a");

// Instantiation
var neuralNetwork = new DN2A.NetworkAlpha();

// Training
// The object passed to the callback function contains information about the training process.
// The list of the valid properties together their accepted ranges and default values is reported in this README file.
var trainingPatterns = [
    {
        input: [0, 0],
        output: [0]
    },
    {
        input: [0, 1],
        output: [1]
    },
    {
        input: [1, 0],
        output: [1]
    },
    {
        input: [1, 1],
        output: [0]
    }
];
neuralNetwork.train(trainingPatterns, function(trainingStatus) {
    console.log("Epoch: " + trainingStatus.elapsedEpochCounter);
});

// Querying
var inputPatterns = [
    [0, 0],
    [0, 1],
    [1, 0],
    [1, 1]
];
neuralNetwork.query(inputPatterns, function(queryingStatus) {
    inputPatterns.forEach(function(inputPatten, inputPatternIndex) {
        console.log("[" + inputPatterns[inputPatternIndex].join(", ") + "] => [" + queryingStatus.outputPatterns[inputPatternIndex].join(", ") + "]");
    });
});
```

### Training and Querying a Single Network through the Cerebrum (ES5) ###

```javascript
// Importation
var DN2A = require("dn2a");

// Instantiation
var cerebrum = new DN2A.Cerebrum({
    minds: [
        {
            name: "firstNeuralNetwork",
            network: {
                generator: DN2A.NetworkAlpha,
                configuration: {
                    layerDimensions: [2, 4, 1],
                    learningMode: "continuous",
                    learningRate: 0.3,
                    momentumRate: 0.7,
                    maximumError: 0.005,
                    maximumEpoch: 1000,
                    dataRepository: {},
                    neuron: {
                        generator: DN2A.Neuron
                    },
                    synapse: {
                        generator: DN2A.Synapse
                    },
                    numbersPrecision: 32
                }
            },
            inputsFrom: [
                "cerebrum"
            ]
        }
    ],
    outputsFrom: [
        "firstNeuralNetwork"
    ]
});

// Training
// The name expected as third parameter by the trainMind method specifies which specific mind to train
var trainingPatterns = [
    {
        input: [0, 0],
        output: [0]
    },
    {
        input: [0, 1],
        output: [1]
    },
    {
        input: [1, 0],
        output: [1]
    },
    {
        input: [1, 1],
        output: [0]
    }
];
cerebrum.trainMind(trainingPatterns, function(trainingStatus) {
    console.log("Epoch: " + trainingStatus.elapsedEpochCounter);
}, "firstNeuralNetwork");

// Querying
// The name expected as third parameter by the queryMind method specifies which specific mind to query
var inputPatterns = [
    [0, 0],
    [0, 1],
    [1, 0],
    [1, 1]
];
cerebrum.queryMind(inputPatterns, function(queryingStatus) {
    inputPatterns.forEach(function(inputPatten, inputPatternIndex) {
        console.log("[" + inputPatterns[inputPatternIndex].join(", ") + "] => [" + queryingStatus.outputPatterns[inputPatternIndex].join(", ") + "]");
    });
}, "firstNeuralNetwork");
```

### Training and Querying an entire Networks Chain  through the Cerebrum (ES5) ###

	TODO

### Training and Querying a Single Network through the Brain (ES5) ###

	TODO

### Training and Querying an entire Networks Chain through the Brain (ES5) ###

	TODO

# Creator #

[Antonio De Luca](http://www.antoniodeluca.info)

----------

# License #

MIT


CHANGELOG.md
## v0.2.1 (6 Sep 2019)

* NPM modules update. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#70](https://github.com/antoniodeluca/dn2a/pull/70)
* Updated README. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#67](https://github.com/antoniodeluca/dn2a/pull/67)
* Neuron generator method tests. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#66](https://github.com/antoniodeluca/dn2a/pull/66)
* Linting up and running. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#64](https://github.com/antoniodeluca/dn2a/pull/64)
* Added eslint and created a first version of the configuration. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#63](https://github.com/antoniodeluca/dn2a/pull/63)
* Included distributed folder into NPM package. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#62](https://github.com/antoniodeluca/dn2a/pull/62)
* Import of used functions only. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#61](https://github.com/antoniodeluca/dn2a/pull/61)
* README corrections. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#60](https://github.com/antoniodeluca/dn2a/pull/60)
* Updated main entry file name, repository url and examples code. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#59](https://github.com/antoniodeluca/dn2a/pull/59)
* Added a CHANGELOG.md file. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#57](https://github.com/antoniodeluca/dn2a/pull/57)
* Removed unwanted badges. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#56](https://github.com/antoniodeluca/dn2a/pull/56)
* Removed not used alias modules. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#55](https://github.com/antoniodeluca/dn2a/pull/55)
* Updated various variable declarations. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#54](https://github.com/antoniodeluca/dn2a/pull/54)
* Removed folder for compiled assets. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#53](https://github.com/antoniodeluca/dn2a/pull/53)
* New building approach, different files structure and fixed bugs. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#52](https://github.com/antoniodeluca/dn2a/pull/52)

## v0.2.0 (1 Nov 2016)

* Fixed version number reported into the Bower configuration. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#49](https://github.com/antoniodeluca/dn2a/pull/49)
* New parameters for the training and querying status objects plus minor changes. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#20](https://github.com/antoniodeluca/dn2a/pull/20)
* Documentation changes and removal of Async. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#19](https://github.com/antoniodeluca/dn2a/pull/19)
* Added initial version of the CONTRIBUTING file needed to clarify the terms under which is possible to contribute to the project. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#18](https://github.com/antoniodeluca/dn2a/pull/18)
* NPM and Bower configurations updates. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#17](https://github.com/antoniodeluca/dn2a/pull/17)
* Improved testing environment configuration and added some badges. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#15](https://github.com/antoniodeluca/dn2a/pull/15)
* Minor improvements and environment configuration for tests. First test containers have been added. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#14](https://github.com/antoniodeluca/dn2a/pull/14)

## v0.1.1 (29 Jul 2016)

* Fixed some typos. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#13](https://github.com/antoniodeluca/dn2a/pull/13)

## v0.1.0 (28 Jul 2016)

* Removed bundle folder from the ignored items of NPM. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#12](https://github.com/antoniodeluca/dn2a/pull/12)
* Neuron and Synapse module refactoring. Minor improvements. Updated documentation. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#11](https://github.com/antoniodeluca/dn2a/pull/11)
* Other improvement to configurability and new README. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#10](https://github.com/antoniodeluca/dn2a/pull/10)
* Improvement to configurability. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#9](https://github.com/antoniodeluca/dn2a/pull/9)
* Decoupled and exposed modules. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#8](https://github.com/antoniodeluca/dn2a/pull/8)
* Added support for momentum. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#7](https://github.com/antoniodeluca/dn2a/pull/7)
* Improvements to the learning algorithm and new examples. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#6](https://github.com/antoniodeluca/dn2a/pull/6)
* Changes for NPM package publication. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#5](https://github.com/antoniodeluca/dn2a/pull/5)
* Fixed problem with tabs and spaces. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#4](https://github.com/antoniodeluca/dn2a/pull/4)
* Added first version of the main documentation file. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#3](https://github.com/antoniodeluca/dn2a/pull/3)
* Completed weights update functionality and improved the whole code. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#2](https://github.com/antoniodeluca/dn2a/pull/2)
* A first implementation!. From [@antoniodeluca](https://github.com/antoniodeluca) with PR [#1](https://github.com/antoniodeluca/dn2a/pull/1)


assets/brain.js
import {
    config
} from "mathjs";

import Cerebrum from "./cerebrum";
import NetworkAlpha from "./networks/alpha";
import Neuron from "./neuron";
import Synapse from "./synapse";

const Brain = function(configuration) {
    this.configuration = configuration || {
        cerebrum: {
            generator: Cerebrum,
            configuration: {
                minds: [
                    {
                        name: "defaultMind",
                        network: {
                            generator: NetworkAlpha,
                            configuration: {
                                layerDimensions: [2, 4, 1],
                                learningMode: "continuous",
                                learningRate: 0.3,
                                momentumRate: 0.7,
                                maximumError: 0.005,
                                maximumEpoch: 1000,
                                dataRepository: {},
                                neuron: {
                                    generator: Neuron
                                },
                                synapse: {
                                    generator: Synapse
                                },
                                numbersPrecision: 32
                            }
                        },
                        inputsFrom: [
                            "cerebrum"
                        ]
                    }
                ],
                outputsFrom: [
                    "defaultMind"
                ]
            }
        }
    };

    if (!this.checkConfiguration()) {
        throw "Invalid Brain Module Configuration";
    }
    this.configuration = this.transformConfiguration();

    config({
        number: "BigNumber",
        precision: this.configuration.numbersPrecision
    });

    this.cerebrum = new this.configuration.cerebrum.generator(this.configuration.cerebrum.configuration);
};

Brain.prototype = {
    checkConfiguration: function() {
        return true;
    },

    transformConfiguration: function() {
        return this.configuration;
    }
};

export default Brain;
